gpustack/worker/serve_manager.py

```
@@ -367,106 +612,52 @@ def _update_model_instance(self, id: int, **kwargs):
-            except NotFoundException:
-                pass
-            # Reset failure counter after transitioning to ERROR
-            self._ready_failures.pop(mi.id, None)
-            return True
-        return False
+        Stop model instance and clean up.
 
-    def _get_model_with_cache(self, mi: ModelInstance) -> Model:
-        """Get model from cache or fetch from clientset."""
-        if mi.id in self._model_cache_by_instance:
-            return self._model_cache_by_instance[mi.id]
+        Args:
+            mi: The model instance to stop.
+        """
 
-        model = self._clientset.models.get(mi.model_id)
-        self._model_cache_by_instance[mi.id] = model
-        return model
+        logger.debug(f"Stopping model instance {mi.name or mi.id}")
 
-    def _stop_model_instance(self, mi: ModelInstance):
-        instance_name = mi.name or f"ModelInstance-{mi.id}"
-        if mi.id not in self._serving_model_instances:
-            logger.warning(f"Model instance {instance_name} is not running. Skipping.")
-            return
-        else:
-            pid = self._serving_model_instances[mi.id].pid
-            try:
-                terminate_process_tree(pid)
-                logger.info(f"Stopped model instance {instance_name} with pid {pid}.")
-            except psutil.NoSuchProcess:
-                logger.warning(
-                    f"Model instance {instance_name} with pid {pid} is already stopped."
-                )
-                pass
-            except Exception as e:
-                logger.error(f"Failed to stop model instance: {pid}: {e}")
+        # Teardown provisioning process if still alive.
+        if self._is_provisioning(mi):
+            terminate_process_tree(self._provisioning_processes[mi.id].pid)
 
-            # For CustomServer backend, also delete the Docker container
-            self._cleanup_inference_server_container(mi)
+        # Delete workload.
+        delete_workload(mi.name)
```

Don't swallow the error of workload deletion, avoid leaving orphan workloadsin host.